File: packages/shared/src/ipc/protocol.ts
// =============================================================================
// JSON-RPC Types
// =============================================================================

export interface JsonRpcRequest {
  jsonrpc: "2.0";
  id: number | string;
  method: string;
  params?: Record<string, unknown>;
}

export interface JsonRpcResponse {
  jsonrpc: "2.0";
  id: number | string;
  result?: unknown;
  error?: { code: number; message: string; data?: unknown };
}

// =============================================================================
// Streaming Response
// =============================================================================

export type StreamResponse =
  | {
      type: "text";
      content: string;
      toolName?: undefined;
      toolArgs?: undefined;
      toolResult?: undefined;
    }
  | {
      type: "code";
      content: string;
      toolName?: undefined;
      toolArgs?: undefined;
      toolResult?: undefined;
    }
  | {
      type: "tool";
      content: string;
      toolName: string;
      toolArgs: unknown;
      toolResult?: string;
    }
  | {
      type: "done";
      content: string;
      toolName?: undefined;
      toolArgs?: undefined;
      toolResult?: undefined;
    }
  | {
      type: "error";
      content: string;
      toolName?: undefined;
      toolArgs?: undefined;
      toolResult?: undefined;
    };

export interface QuestionSpec {
  question: string;
  header?: string;
  options: Array<{ label: string; description: string }>;
  multiple?: boolean;
  custom?: boolean;
}

/** User's answer to a permission prompt */
export type PermissionPromptDecision =
  | "allow-once"
  | "allow-session"
  | "allow-project"
  | "allow-always"
  | "deny";

// `sessionId` is populated on every variant when the event is relayed through
// the bus (see `busEventToClientEvent` in apps/core/src/bus/bridge.ts, which
// stamps it from the `StreamRelayEvent` wrapper) — it's optional here because
// event *authors* below construct these payloads without it. Consumers that
// multiplex several sessions over one process's stdout (e.g. frontends
// driving multiple concurrent sessions) need it to route each line; the TUI
// and per-session SSE channel don't, since they're already scoped to one
// session and can ignore the field.
export type StreamEvent =
  | {
      type: "tool_start";
      sessionId?: string;
      toolCallId: string;
      toolName: string;
      args: Record<string, unknown>;
    }
  | { type: "tool_output"; sessionId?: string; toolCallId: string; content: string }
  | {
      type: "tool_complete";
      sessionId?: string;
      toolCallId: string;
      toolName: string;
      result: string;
      success: boolean;
      duration_ms?: number;
    }
  | { type: "thinking"; sessionId?: string; content: string } // full thinking, emitted at turn end (non-stream path)
  | { type: "text"; sessionId?: string; content: string } // full assistant text, emitted at turn end (non-stream path or as compatibility snapshot when streaming)
  | { type: "text_delta"; sessionId?: string; delta: string } // incremental assistant text chunk (streaming path)
  | { type: "thinking_delta"; sessionId?: string; delta: string } // incremental reasoning chunk (streaming path)
  | { type: "done"; sessionId?: string; content: string }
  | { type: "error"; sessionId?: string; content: string }
  // Follow-up queue (spec 2026-08-05-queued-messages-design): a session.send
  // arrived while a turn was already in progress and was parked instead of
  // racing. `id` matches the `QueuedMessage`; `content` is the original prompt
  // so the UI can echo it back. `message_dequeued` fires when the same id is
  // pulled out via session.dequeue (or, implicitly, when the queue finally
  // starts its turn — the UI treats that as a state change to "in-flight").
  | { type: "message_queued"; sessionId?: string; id: string; content: string }
  | { type: "message_dequeued"; sessionId?: string; id: string }
  // A memory was written about the user WITHOUT them asking (turn-end
  // extraction, spec 2026-08-09-memory-write-path D5). Arrives after the
  // turn's `done` because extraction is fire-and-forget, so frontends must
  // treat it as an out-of-band notice rather than part of the turn. Never
  // emitted for the `memory` tool — that already shows as a tool call.
  | {
      type: "memory_saved";
      sessionId?: string;
      memories: Array<{ type: string; name: string }>;
    }
  // Automatic (non-tool-call) retrieval surfaced one or more saved memories
  // into this turn's prompt (spec D5, graph/index.ts prepareMemories).
  // Emitted once per distinct user message that gets a hit — not every
  // inner-loop tool-call turn — so the UI shows a single notice per request
  // rather than repeating it. Silent when nothing relevant was found.
  | {
      type: "memory_injected";
      sessionId?: string;
      memories: Array<{ type: string; name: string }>;
    }
  | {
      // Prompt-cache awareness (jcode #9). "cold" is emitted before a send that
      // will likely miss Anthropic's ~5-min cache; "warm" carries post-turn
      // read/write token counts. Informational — frontends may render or ignore.
      type: "cache_status";
      sessionId?: string;
      // "miss" is the harness-bug alarm (spec 2026-08-09 D2): the prefix was
      // busted with no recorded cause, i.e. something changed a message that
      // had already been sent. A legitimate rebuild (compaction) is documented
      // in the invalidation journal and never reaches the frontend.
      state: "cold" | "warm" | "miss";
      message?: string; // human-readable, set on "cold" and "miss"
      cacheReadTokens?: number; // served from cache (cheap), set on "warm"/"miss"
      cacheWriteTokens?: number; // written to cache this turn, set on "warm"/"miss"
    }
  // Turn-level advisory the user needs to see (e.g. an attachment dropped
  // because the model can't accept it). Does not fail the turn.
  | { type: "notice"; sessionId?: string; level: "info" | "warn"; content: string }
  // Running spend for the current run() — emitted once per completed turn so
  // the frontend can render a live per-session counter (spec
  // 2026-08-05-token-efficiency, D7). Core computes; frontends only display.
  | {
      type: "usage_totals";
      sessionId?: string;
      // Already includes cache writes — they are billed as input. The separate
      // totalCacheWriteTokens below is the same tokens broken out for the cache
      // hit rate, so summing the two double-counts them.
      totalInputTokens: number;
      totalOutputTokens: number;
      totalCacheReadTokens: number;
      totalCacheWriteTokens?: number;
    }
  | { type: "compaction_start"; sessionId?: string; trigger: "auto" | "manual" } // compaction began
  | {
      type: "compaction_complete";
      sessionId?: string;
      trigger: "auto" | "manual";
      compacted: boolean; // false when there was nothing to compact / it was blocked
      tokensBefore: number;
      tokensAfter: number;
      reason?: string; // why it didn't compact, when compacted=false
    }
  | {
      type: "question_asked";
      requestId: string;
      sessionId?: string;
      questions: QuestionSpec[];
    }
  | {
      type: "permission_asked";
      requestId: string;
      sessionId?: string;
      toolName: string;
      args: Record<string, unknown>;
      /** Human-readable summary, e.g. the bash command or file path */
      description: string;
      /** Rule offered for "always allow", e.g. "Bash(npm run test:*)" */
      suggestedRule?: string;
      /** Which rule or mode default triggered the ask */
      reason?: string;
    };

// =============================================================================
// IPC Method Signatures
// =============================================================================

export const METHODS = {
  "tools.list": {
    params: undefined,
    result: [] as import("../types.js").ToolListItem[],
  },
  "tools.call": {
    params: { name: "", args: {} as Record<string, unknown> },
    result: {} as import("../types.js").ToolResult,
  },
  "session.start": {
    params: { projectPath: "", provider: "" },
    result: { sessionId: "" },
  },
  "session.send": {
    // METHODS entries are values, not types — the exported MethodParams<M>
    // reads their inferred shape, so an optional field needs a cast.
    params: {} as {
      sessionId: string;
      message: string;
      images?: Array<{ data: string; mediaType: string; altText?: string }>;
    },
    // The LoopResult shape when the turn ran normally. When the session was
    // already busy, the call parks the prompt in the follow-up queue and
    // resolves immediately with { queued: true, id } — the UI uses the
    // `message_queued` stream event for the same data so web/SSE subscribes
    // stay in sync.
    result: {} as
      | StreamResponse
      | { queued: true; id: string },
  },
  "session.dequeue": {
    // Removes a previously-queued message by id (no-op if it already started
    // sending). The result tells the caller whether something was actually
    // removed, so the UI can decide between "drop the indicator" and "ignore
    // — the message is already in flight".
    params: { sessionId: "" as string, id: "" as string },
    result: { removed: false as boolean },
  },
  "session.stop": {
    params: { sessionId: "" },
    result: undefined as void,
  },
  "session.compact": {
    params: { sessionId: "" },
    result: {} as {
      compacted: boolean;
      tokensBefore: number;
      tokensAfter: number;
      reason?: string;
    },
  },
  "session.list": {
    // METHODS entries are values, not types — an optional field needs a cast.
    params: {} as {
      projectPath?: string;
      status?: import("../types.js").SessionStatus;
    },
    result: [] as import("../types.js").SessionMeta[],
  },
  "session.resume": {
    params: { sessionId: "" },
    result: {} as import("../types.js").SessionResumeResult,
  },
  // Lists Claude Code sessions discovered under $CLAUDE_CONFIG_DIR (defaults
  // to ~/.claude). Read-only: core never writes to the user's Claude Code
  // store. See apps/core/src/claude-sessions/.
  "session.claudeList": {
    params: {} as {
      projectPath?: string;
      limit?: number;
    },
    result: [] as import("../types.js").ClaudeSessionMeta[],
  },
  // Reads a Claude Code session transcript and converts it to the
  // SerializedMessage shape so the frontend can reuse the existing
  // preview-markdown renderer.
  "session.claudeTranscript": {
    params: { sessionId: "" },
    result: {} as import("../types.js").ClaudeTranscript,
  },
  "providers.list": {
    params: undefined,
    result: [] as import("../types.js").ProviderInfo[],
  },
  "commands.list": {
    params: { projectPath: "" },
    result: [] as import("../types.js").CommandInfo[],
  },
  "commands.resolve": {
    params: { name: "", args: [] as string[], projectPath: "" },
    result: { prompt: "" },
  },
  "question.answer": {
    params: { requestId: "", answers: [] as string[] },
    result: undefined as void,
  },
  "question.reject": {
    params: { requestId: "" },
    result: undefined as void,
  },
  "permission.answer": {
    params: {
      requestId: "",
      decision: "deny" as PermissionPromptDecision,
      editedRule: undefined as string | undefined,
    },
    result: undefined as void,
  },
  "permission.reject": {
    params: { requestId: "" },
    result: undefined as void,
  },
  "usage.get": {
    params: undefined,
    result: [] as { date: string; tokencount: number }[],
  },
  "skills.list": {
    params: { projectPath: undefined as string | undefined },
    result: [] as {
      name: string;
      description?: string;
      scope: string;
    }[],
  },
  "mcp.status": {
    params: { name: undefined as string | undefined },
    result: [] as {
      name: string;
      type: string;
      enabled: boolean;
      status: "connected" | "disconnected";
      toolCount: number;
      tools: string[];
    }[],
  },
  // Persisted prompt history — up-arrow recall across sessions. Core owns
  // ~/.freecode/history.jsonl; the TUI seeds its in-memory ring at startup
  // and appends on every submit.
  "history.list": {
    params: undefined,
    result: [] as string[],
  },
  "history.append": {
    params: { text: "" },
    result: undefined as void,
  },
  // Open the optional graph explorer in the browser. Returns { url } on
  // success; { error: "not-installed" } if the user hasn't run
  // `freecode memory ui-install` yet (the addon is opt-in, ~280 KB download
  // from the GitHub release). Spec: 2026-08-04-memory-graph-explorer-design.md.
  "graph.explore": {
    params: undefined,
    result: {} as { url: string } | { error: "not-installed" },
  },
} as const;

export type MethodName = keyof typeof METHODS;
export type MethodParams<M extends MethodName> = (typeof METHODS)[M]["params"];
export type MethodResult<M extends MethodName> = (typeof METHODS)[M]["result"];

File: apps/tui-rs/src/ipc/protocol.rs
//! Mirrors packages/shared/src/ipc/protocol.ts and types.ts.
//! Keep in sync with the TS source of truth when the wire protocol changes.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Serialize)]
pub struct JsonRpcRequest {
    pub jsonrpc: &'static str,
    pub id: u64,
    pub method: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub params: Option<Value>,
}

#[derive(Debug, Deserialize)]
pub struct JsonRpcError {
    #[allow(dead_code)]
    pub code: i64,
    pub message: String,
}

#[derive(Debug, Deserialize)]
pub struct JsonRpcResponse {
    pub id: Value,
    #[serde(default)]
    pub result: Option<Value>,
    #[serde(default)]
    pub error: Option<JsonRpcError>,
}

/// Streamed while a turn is in flight. Lines on stdout that carry a "type"
/// but no "jsonrpc" field are StreamEvents rather than JSON-RPC responses.
#[derive(Debug, Clone, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum StreamEvent {
    ToolStart {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        args: Value,
    },
    ToolOutput {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        content: String,
    },
    ToolComplete {
        #[serde(rename = "toolCallId")]
        tool_call_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        result: String,
        success: bool,
        #[serde(rename = "duration_ms")]
        duration_ms: Option<u64>,
    },
    Thinking {
        content: String,
    },
    Text {
        content: String,
    },
    TextDelta {
        delta: String,
    },
    ThinkingDelta {
        delta: String,
    },
    Done {
        content: String,
    },
    Error {
        content: String,
    },
    /// Compaction began (summarizing older turns to free context).
    CompactionStart {
        #[serde(default)]
        trigger: String,
    },
    /// Compaction finished; carries the before/after token estimates.
    CompactionComplete {
        #[serde(default)]
        trigger: String,
        #[serde(default)]
        compacted: bool,
        #[serde(rename = "tokensBefore", default)]
        tokens_before: u64,
        #[serde(rename = "tokensAfter", default)]
        tokens_after: u64,
        #[serde(default)]
        reason: Option<String>,
    },
    /// Core is blocked waiting for the user to answer. Ignoring these hangs
    /// the turn — the agent loop only resumes on `question.answer`/`reject`.
    QuestionAsked {
        #[serde(rename = "requestId")]
        request_id: String,
        questions: Vec<QuestionSpec>,
    },
    /// Core is blocked waiting for a permission decision (`permission.answer`).
    PermissionAsked {
        #[serde(rename = "requestId")]
        request_id: String,
        #[serde(rename = "toolName")]
        tool_name: String,
        #[serde(default)]
        description: String,
        #[serde(rename = "suggestedRule", default)]
        suggested_rule: Option<String>,
        #[serde(default)]
        reason: Option<String>,
    },
    /// Synthesized locally (core never sends this) from the `session.send`
    /// RPC result's usage, so the status bar's context meter can update.
    Usage {
        context_tokens: u64,
    },
    /// Synthesized locally after a model switch (core never sends this) so the
    /// status meter's denominator tracks the newly selected model's window.
    ContextLimit {
        limit: u64,
    },
}

/// One question in a `question_asked` event; mirrors `QuestionSpec` in
/// `packages/shared/src/ipc/protocol.ts`.
#[derive(Debug, Clone, Deserialize)]
pub struct QuestionSpec {
    pub question: String,
    #[serde(default)]
    pub header: Option<String>,
    #[serde(default)]
    pub options: Vec<QuestionOption>,
    #[serde(default)]
    pub multiple: bool,
}

#[derive(Debug, Clone, Deserialize)]
pub struct QuestionOption {
    pub label: String,
    #[serde(default)]
    pub description: String,
}

/// A line read from the core's stdout is either a JSON-RPC response
/// (has "jsonrpc") or a bare StreamEvent (has "type" only).
pub enum IncomingLine {
    Response(JsonRpcResponse),
    Event(StreamEvent),
    Unparseable,
}

pub fn parse_line(line: &str) -> IncomingLine {
    let trimmed = line.trim();
    if trimmed.is_empty() {
        return IncomingLine::Unparseable;
    }
    let Ok(value) = serde_json::from_str::<Value>(trimmed) else {
        return IncomingLine::Unparseable;
    };
    if value.get("jsonrpc").is_some() {
        return match serde_json::from_value::<JsonRpcResponse>(value) {
            Ok(resp) => IncomingLine::Response(resp),
            Err(_) => IncomingLine::Unparseable,
        };
    }
    if value.get("type").is_some() {
        return match serde_json::from_value::<StreamEvent>(value) {
            Ok(evt) => IncomingLine::Event(evt),
            Err(_) => IncomingLine::Unparseable,
        };
    }
    IncomingLine::Unparseable
}

#[derive(Debug, Clone, Serialize)]
pub struct SessionConfig {
    #[serde(rename = "projectPath")]
    pub project_path: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider: Option<String>,
    #[serde(rename = "agentMode", skip_serializing_if = "Option::is_none")]
    pub agent_mode: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct SessionInfo {
    #[serde(rename = "sessionId")]
    pub session_id: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ProviderInfo {
    pub id: String,
    pub name: String,
    #[serde(default)]
    pub description: String,
    /// Whether core can find an API key for this provider (config or env). The
    /// picker badges providers that still need one.
    #[serde(rename = "hasApiKey", default)]
    pub has_api_key: bool,
}

#[derive(Debug, Clone, Default, Deserialize)]
pub struct SessionSendResult {
    #[allow(dead_code)]
    pub success: bool,
    #[serde(default)]
    pub content: Option<String>,
    #[serde(default)]
    pub usage: Option<TokenUsage>,
}

/// Token accounting from the agent loop's final `LoopResult.usage`
/// (`apps/core/src/agent/types.ts`). `context_tokens` is the last API call's
/// full input — true context-window occupancy; when the core omits it we fall
/// back to `input + cache_read`, mirroring `apps/tui/src/index.ts`.
#[derive(Debug, Clone, Deserialize)]
pub struct TokenUsage {
    #[serde(rename = "inputTokens", default)]
    pub input_tokens: u64,
    #[serde(rename = "cacheReadInputTokens", default)]
    pub cache_read_input_tokens: u64,
    #[serde(rename = "contextTokens", default)]
    pub context_tokens: Option<u64>,
}

impl TokenUsage {
    /// True context-window occupancy: the core's `contextTokens` when present,
    /// otherwise the input + cache-read fallback the TS TUI uses.
    pub fn context_tokens(&self) -> u64 {
        self.context_tokens
            .unwrap_or(self.input_tokens + self.cache_read_input_tokens)
    }
}

#[derive(Debug, Clone, Deserialize)]
pub struct ToolListItem {
    #[allow(dead_code)]
    pub id: String,
    #[allow(dead_code)]
    pub description: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct CurrentModel {
    pub provider: String,
    pub model: String,
}

/// One entry from `models.list`; mirrors `ModelInfo` in
/// `apps/tui/src/ipc/client.ts`. Only `id` is required — `name` falls back to
/// the id, and the model's context/output limits are ignored here (the status
/// bar resolves the window via `models.contextLimit`).
#[derive(Debug, Clone, Deserialize)]
pub struct ModelInfo {
    pub id: String,
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
}
